# <div align="center"> <img src="https://raw.githubusercontent.com/JessicaRocano/IA/master/files/ups-logo.png">
# <div align="center">  PROYECTO INTERCICLO
    
    
    
    
### <div align="center"> MATERIA:
### <div align="center"> Inteligencia Artificial 1 & Sistemas Expertos
    
    
## <div align="center"> DOCENTE:
### <div align="center"> Ing. Diego Quisi
## <div align="center"> 
## <div align="center">INTEGRANTES
#### <div align="center"> Jessica Rocano
#### <div align="center"> Carlos Andrade
 ## </div>
## </div>
    
   
***
- - -


# <div align="center"> Algoritmos  en Neo4j

La biblioteca Neo4j Graph Data Science (GDS) contiene muchos algoritmos gráficos. Los algoritmos se dividen en categorías que representan diferentes clases de problemas.
   

### <div align="center"> Algoritmo de Centralidad </div> ###
 Los algoritmos de centralidad se utilizan para determinar la importancia de distintos nodos en una red de datos.
 ## <div align="left"> Page Rank (Rango de pagina)</div>
 - ### INTRODUCCION:
El algoritmo de Page Rank mide la importancia de cada nodo dentro del grafo, en función del número de relaciones entrantes y la importancia de los nodos de origen correspondientes. Una página es tan importante como las páginas que enlazan con ella.
  
 - ### DESCRIPCION:
 Page Rank se define en el documento original de Google como una función que resuelve la siguiente ecuación:
 PR(A) = (1-d) + d (PR(T1)/C(T1) + ... + PR(Tn)/C(Tn))
 
 DONDE:
   * suponemos que una página Atiene páginas T1 a las Tn que apuntan (es decir, son citas).
   * d es un factor de amortiguación que se puede establecer entre 0 y 1. Por lo general, se establece en 0,85.
   * C(A)se define como el número de enlaces que salen de la página A.
   
   Esta ecuación se usa para actualizar iterativamente una solución candidata y llegar a una solución aproximada a la misma ecuación.
   
   La ejecución de este algoritmo requiere suficiente disponibilidad de memoria.
   
   
- ### PROCESO:
    * Sintaxis: 
      - #### Modo de escritura: ####
          PageRank en modo de escritura en un gráfico almacenado en el catálogo.
          ~~~
          CALL gds.pageRank.write(
          graphName: String,
          configuration: Map
            )
            YIELD
          // general write return columns
          ranIterations: Integer,
          didConverge: Boolean
          ~~~
              donde:
              graphName: representa el nombre de un gráfico almacenado en el catálogo.
              configuration: reprenta la configuración para algoritmos específicos y / o filtrado de gráficos.
       - #### Configuración específica: ####
           se debe especificar los siguientes parámetros
            - dampingFactor: el valor de 0.85 por defecto, factor de amortiguación del cálculo del rango de página.
            - maxIterations: el valor de 20 por defecto, represnta número máximo de iteraciones de Page Rank para ejecutar.
            - tolerance: el valor de 0.0000001 por defecto, representa el cambio mínimo en las puntuaciones entre iteraciones. Si todas las puntuaciones cambian menos que el valor de tolerancia, el resultado se considera estable y el algoritmo regresa.
            - relationshipWeightProperty. el valor por defecto es null, de tipo opcional , permite especificar el nombre de la propiedad que contiene peso. Si es nulo, trata el gráfico como no ponderado. Debe ser numérico.

- ### EJEMPLO SENCILLO:



       


In [5]:
from py2neo import Graph
graph = Graph(password="jessica")

In [6]:
from py2neo import Node, Relationship

In [20]:
alice = Node("Person", name="Alice")
bob = Node("Person", name="Bob")
alice_knows_bob = Relationship(alice, "KNOWS", bob)
graph.create(alice_knows_bob)

In [ ]:
prueba crear nodo

In [30]:
graph.run("CREATE (Paco:Person {name:'Andre', born:1964})")

DATOS INICIALES

In [7]:
for record in graph.run("MATCH(n) RETURN n AS name"):
    print(record['name'])

(_0:Colegio {Ciudad: 'Cuenca', name: 'Colegio Militar Abdon Calderon', valor: '3.9'})
(_1:Colegio {Ciudad: 'Cuenca', name: 'Colegio Benigno Malo', valor: '4.6'})
(_2:Colegio {Ciudad: 'Cuenca', name: 'Colegio Tecnico Salesiano', valor: '4.5'})
(_3:Colegio {Ciudad: 'Cuenca', name: 'Colegio interandino', valor: '3.5'})
(_4:Colegio {Ciudad: 'Cuenca', name: 'Colegio Continental', valor: '3.4'})
(_5:Colegio {Ciudad: 'Cuenca', name: 'Unidad Educativa Los Andes', valor: '4.5'})
(_6:Colegio {Ciudad: 'Cuenca', name: 'Colegio De Bachillerato T\u00e9cnico Daniel C\u00f3rdova Toral', valor: '5.0'})
(_7:Colegio {Ciudad: 'Cuenca', name: 'Colegio Herlinda Toral', valor: '4.5'})
(_8:Colegio {Ciudad: 'Cuenca', name: 'Unidad Educativa "Hermano Miguel " LA SALLE.', valor: '4.4'})
(_9:Colegio {Ciudad: 'Cuenca', name: 'Unidad Educativa Particular Dominicana "San Luis Beltr\u00e1n"', valor: '5.0'})
(_10:Hospital {Ciudad: 'Cuenca', name: 'Hospital Regional Jos\u00e9 Carrasco Arteaga (IESS)', valor: '3.5'})
(_

## PAGE RANK ##
### ejemplo con datos reales ###
#### Descripcion del problema: ####

Se desea obtener recomendaciones de centros hospitalarios para una persona que requiera de atención médica, teniendo en cuenta la cantidad y calidad de las conexiones que apunten al lugar de partida. 


#### creará el gráfico y lo almacenará en el catálogo de gráficos.##
~~~~
CALL gds.graph.create(
    'myGraph',
    'Page',
    'LINKS',
    {
        relationshipProperties: 'weight'
    }
)
~~~~

##### Lo siguiente ejecutará el algoritmo y los resultados de la transmisión: #####

In [12]:
query = """
CALL gds.pageRank.stream('myGraph', { maxIterations: 20, dampingFactor: 0.85 })
YIELD nodeId, score
RETURN gds.util.asNode(nodeId).name AS name, score
ORDER BY score DESC, name ASC

"""

data = graph.run(query)

for d in data:
    print(d)

<Record name='Hospital Regional José Carrasco Arteaga (IESS)' score=0.5058884598314762>
<Record name='IESS - Centro de Salud C - Materno Infantil y Emergencias – Cuenca' score=0.4186923070345074>
<Record name='Hospital de Atención Integral al Adulto Mayor' score=0.3596109995851293>
<Record name='Hospital IESS Santo Domingo de los Tsachilas' score=0.3573814126197249>
<Record name='Hospital Dr. Gustavo Dominguez Zambrano' score=0.33168750405311587>
<Record name='Hospital Vicente Corral Moscoso' score=0.3161085980944336>
<Record name='Hospital Del iess (La Cadena)' score=0.31362500544637445>
<Record name='Centro De Salud' score=0.30459375455975535>
<Record name='Hospital Esperanza' score=0.30459375455975535>
<Record name='Hospital IESS de Machala' score=0.30459375455975535>
<Record name='Hospital Abel Gilbert Pontón (Hospital Guayaquil)' score=0.2832906267605722>
<Record name='Hospital de Especialidades Teodoro Maldonado Carbo' score=0.2832906267605722>
<Record name='Clinica Niño Jesus' s

###### Estimación de memoria######
Lo siguiente estimará los requisitos de memoria para ejecutar el algoritmo: 

In [10]:
query = """
CALL gds.pageRank.write.estimate('myGraph', {
  writeProperty: 'pageRank',
  maxIterations: 20,
  dampingFactor: 0.85
})
YIELD nodeCount, relationshipCount, bytesMin, bytesMax, requiredMemory
"""

data = graph.run(query)

for d in data:
    print(d)

<Record nodeCount=50 relationshipCount=45 bytesMin=3032 bytesMax=3032 requiredMemory='3032 Bytes'>


#####  Lo siguiente ejecutará el algoritmo y devolverá el resultado en forma de valores estadísticos y de medición ######

In [13]:
query = """
CALL gds.pageRank.stats('myGraph', {
  maxIterations: 20,
  dampingFactor: 0.85
})
YIELD centralityDistribution
RETURN centralityDistribution.max AS max
"""

data = graph.run(query)

for d in data:
    print(d)

<Record max=0.5058889389038086>


La consulta devolvió un llamado histograma de centralidad que puede usarse para monitorear la distribución de los valores del camino en todos los nodos calculados.

## Strongly Connected Components ##
### ejemplo con datos reales ###
#### Descripcion del problema: ####

Se desea calcular la conectividad entre diferentes parques de una ciudad para poder estimar el rendimiento de enrutamiento en redes inalámbricas entre múltiples parques de la ciudad de Quito.  


##### Lo siguiente ejecutará el algoritmo y transmitirá los resultados:  #####

In [28]:
query = """
CALL gds.alpha.scc.stream({
  nodeQuery: 'MATCH (u:Parque) where u.Ciudad="Quito"  RETURN id(u)  AS id',
relationshipQuery: 'MATCH (u1:Parque)-[:DISTANCIA]->(u2:Parque) where u1.Ciudad="Quito" AND u2.Ciudad="Quito" RETURN id(u1) AS source, id(u2) AS target' })
YIELD nodeId, componentId
RETURN gds.util.asNode(nodeId).name AS Name, componentId AS Component
ORDER BY Component DESC

"""

data = graph.run(query)

for d in data:
    print(d)

<Record Name='Parque El Ejido' Component=9>
<Record Name='Parque Metropolitano del Sur' Component=9>
<Record Name='Parque La Alameda' Component=9>
<Record Name='Parque Itchimbía' Component=9>
<Record Name='Parque Metropolitano Equinoccial' Component=9>
<Record Name='Parque Guápulo' Component=9>
<Record Name='Parque Puertas del Sol' Component=9>
<Record Name='Parque Bicentenario' Component=8>
<Record Name='Parque Arqueológico Rumipamba' Component=6>
<Record Name='Parque Metropolitano Guangüiltagua' Component=0>


In [ ]:
podemos observar que el componente altamente conectado es Parque Metropolitano Guangüiltagua seguido de Parque Arqueológico Rumipamba

## similitud de Jackard ##
### ejemplo con datos reales ###
#### Descripcion del problema: ####
se necesita obtener una recomendación de rutas a seguir de acuerdo a las similitudes que se hallen entre los caminos cercanos  a la iglesia  El Belén perteneciente a la ciudad de Quito.  




In [ ]:
Lo siguiente devolverá la similitud de la iglesia  El Belén  y las otras Iglesias que tienen una discancia en Comun

In [30]:
query = """
MATCH (p1:Iglesia {name: 'Iglesia El Belén'})-[:DISTANCIA]->(cuisine1)
WITH p1, collect(id(cuisine1)) AS p1Cuisine
MATCH (p2:Iglesia)-[:DISTANCIA]->(cuisine2) WHERE p1 <> p2
WITH p1, p1Cuisine, p2, collect(id(cuisine2)) AS p2Cuisine
RETURN p1.name AS from,
       p2.name AS to,
       gds.alpha.similarity.jaccard(p1Cuisine, p2Cuisine) AS similarity
ORDER BY to, similarity DESC


"""

data = graph.run(query)

for d in data:
    print(d)

<Record from='Iglesia El Belén' to='Basílica Católica de San Pedro de Conocoto' similarity=0.0>
<Record from='Iglesia El Belén' to='Catedral Católica de Machala' similarity=0.0>
<Record from='Iglesia El Belén' to='Catedral De Cuenca' similarity=0.0>
<Record from='Iglesia El Belén' to='Catedral Metropolitana De Guayaquil' similarity=0.0>
<Record from='Iglesia El Belén' to='Catedral Metropolitana de Quito' similarity=0.3333333333333333>
<Record from='Iglesia El Belén' to='Catedral Vieja de Cuenca' similarity=0.0>
<Record from='Iglesia El Belén' to='El Señor de los Puentes' similarity=0.0>
<Record from='Iglesia El Belén' to='Iglesia Bíblica Bautista "Rey de Vida"' similarity=0.0>
<Record from='Iglesia El Belén' to='Iglesia Católica "Nuestra Señora de Chilla" - Machala' similarity=0.0>
<Record from='Iglesia El Belén' to='Iglesia Católica Cristo del Consuelo - Machala' similarity=0.0>
<Record from='Iglesia El Belén' to='Iglesia Católica Jesús El Buen Pastor' similarity=0.0>
<Record from='Ig

Se puede observar que la Iglesia El Belén a la iglesia Catedral Metropolitana de Quito tiene una  similaritud de 0.3333333333333333 mientras que la Iglesia de Cristo la Colón tiene una similitud de 0.25.
la misma que nos da a entender que el punto mas cercano es la Iglesia de Cristo la Colón

In [ ]:
## similitud de Jackard ##
### ejemplo con datos reales ###
#### Descripcion del problema: ####
Se desea calcular la cercanía de los diferentes puntos de centros educativos  tomando como principal referencia la función de sus vecinos compartidos. La misma que nos ofrezca una medida con la que se pueda estimar con precisión la cercanía entre sus vértices pertenecientes a un grafo.



Lo siguiente devolverá el puntaje de asignación de recursos para 'Unidad Educativa Los Andes' y Colegio Tecnico Salesiano: 

In [31]:
query = """
MATCH (p1:Colegio {name: 'Unidad Educativa Los Andes'})
MATCH (p2:Colegio {name: 'Colegio Tecnico Salesiano'})
 RETURN gds.alpha.linkprediction.resourceAllocation(p1, p2, {relationshipQuery: "DISTANCIA"}) AS score

"""

data = graph.run(query)

for d in data:
    print(d)

<Record score=0.25>


In [ ]:
Dentro de los resultados que podemos obtener con este algoritmo se tiene que los valores 0 indican que los nodos que son objeto de estudio no están cerca, mientras que en el caso de valores altos indican que los nodos están más cerca.

##  ##
### ejemplo con datos reales ###
#### Descripcion del problema: ####
se necesita obtener una recomendación de mercados a visitar con menor distancia posible, que el mismo nos permita ahorrar tiempo y esfuerzo.



In [ ]:
Lo siguiente ejecutará el algoritmo y los resultados de la transmisión: 

In [32]:
query = """
MATCH (n:Mercado {name: 'Mercado de San Roque'}) where n.Ciudad="Quito" 
CALL gds.alpha.shortestPath.deltaStepping.stream({
  nodeProjection: 'Mercado',
  relationshipProjection: {
    DISTANCIA: {
      type: 'DISTANCIA',
      properties: 'weight'
    }
  },
  startNode: n,
  relationshipWeightProperty: 'weight',
  delta: 3.0
})
YIELD nodeId, distance
RETURN gds.util.asNode(nodeId).name AS destination, distance
"""

data = graph.run(query)

for d in data:
    print(d)

<Record destination='Mercado Sur Machala' distance=inf>
<Record destination='Mercado Central De Machala' distance=inf>
<Record destination='Mercado Municipal Pto. Bolívar' distance=inf>
<Record destination='Feria libre' distance=inf>
<Record destination='Supermercados del Norte' distance=inf>
<Record destination='Supermaxi Machala' distance=inf>
<Record destination='Mercado Municipal De Mayoristas Y Minoristas De Machala.' distance=inf>
<Record destination='AKÍ Machala' distance=inf>
<Record destination='TÍA Machala' distance=inf>
<Record destination='Super AKÍ Palmeras' distance=inf>
<Record destination='Mercado Mapasingue Oeste' distance=inf>
<Record destination='Mercado Artesanal' distance=inf>
<Record destination='Mercado informal Paraiso de la Flor' distance=inf>
<Record destination='Mercado Municipal "Bastión Popular"' distance=inf>
<Record destination='Mercado Transferencia de Viveres' distance=inf>
<Record destination='Mercado Oeste' distance=inf>
<Record destination='Mercado J

In [ ]:
El resultado indicado en la tabla nos dice que partiendo desde Mercado de San Roque hacia otro nodo el costo de operación es cero, es decir, la ruta más corta, es la que inicia desde ese vértice.